File: src/lib/relevance.ts
// ─────────────────────────────────────────────────────────────────────────────
// relevance.ts  —  Port of compute_relevance_scores() + corpus build (CELL 4)
//
// Level 2 contrastive bi-encoder relevance (BAAI/bge-base-en-v1.5 via the HF
// router). Hard rule preserved: NO TF-IDF fallback. Any HF failure throws
// HFAPIError so the UI surfaces a "Try again" state — it never silently degrades.
//
// Logic preserved verbatim:
//   • Surgical corpus: query/title/H1 + topic-bearing H2/meta fragments only.
//   • Dynamic collision check drops hard negatives that share a topic token.
//   • Contrastive query = topic_emb − 0.15 · mean(active_negative_embs), renormed.
//   • Query-only matches use the escalated 6.5 floor; others use 5.0. Below → 0.
// ─────────────────────────────────────────────────────────────────────────────
import {
  HF_API_TOKEN, HF_EMBED_MODEL, HARD_NEGATIVES, _DRIFT_NEGATIVES, CONTRASTIVE_WEIGHT,
  NLP_BATCH_SIZE, NLP_EMBED_WORKERS, RELEVANCE_MIN_THRESHOLD, QUERY_ONLY_MIN_THRESHOLD,
  SCORE_MIN, SCORE_MAX,
} from "./config";
import {
  CandidateRow, LogFn, l2normalize, cosineToQuery, clampRound1, mapWithConcurrency, sleep,
} from "./util";

export class HFAPIError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "HFAPIError";
  }
}

const GENERIC_TOKENS = new Set([
  "software", "platform", "platforms", "tool", "tools", "service", "services",
  "solution", "solutions", "system", "systems", "app", "apps", "online",
  "best", "top", "company", "companies", "provider", "providers", "management",
]);
const MATCH_STOPWORDS = new Set(["the", "and", "for", "with", "your", "you", "our", "from", "that", "this", "are"]);

const EMBED_URL = `https://router.huggingface.co/hf-inference/models/${HF_EMBED_MODEL}/pipeline/feature-extraction`;

async function postEmbed(texts: string[], timeoutMs: number): Promise<Response> {
  const ctrl = new AbortController();
  const t = setTimeout(() => ctrl.abort(), timeoutMs);
  try {
    return await fetch(EMBED_URL, {
      method: "POST",
      headers: { Authorization: `Bearer ${HF_API_TOKEN}`, "Content-Type": "application/json" },
      body: JSON.stringify({ inputs: texts }),
      signal: ctrl.signal,
    });
  } finally {
    clearTimeout(t);
  }
}

export async function warmupEmbedModel(): Promise<void> {
  try {
    await postEmbed(["warmup"], 30000);
  } catch {
    /* fire and forget */
  }
}

function buildNlpCorpus(rows: CandidateRow[], rawTopic: string): string[] {
  const topicTerms = (rawTopic.toLowerCase().match(/[a-z0-9]+/g) ?? []).filter((t) => t.length > 2);
  return rows.map((r) => {
    const bits = [String(r.query ?? r.query_all ?? ""), String(r.meta_title ?? ""), String(r.h1 ?? "")];
    for (const col of [r.h2 ?? "", r.meta_description ?? ""]) {
      const val = String(col);
      if (!val) continue;
      const fragments = val.split(/[|•♦\n.\-]/).map((f) => f.trim()).filter(Boolean);
      for (const frag of fragments) {
        if (topicTerms.some((term) => frag.toLowerCase().includes(term))) bits.push(frag);
      }
    }
    return bits.join(" ");
  });
}

export async function computeRelevanceScores(
  rows: CandidateRow[],
  rawTopic: string,
  log: LogFn = console.log
): Promise<number[]> {
  if (!rows.length) return [];
  if (!HF_API_TOKEN || !HF_API_TOKEN.startsWith("hf_")) {
    throw new HFAPIError("HF_API_TOKEN is missing or invalid.");
  }

  const corpus = buildNlpCorpus(rows, rawTopic);

  // ── Dynamic collision check ────────────────────────────────────────────────
  const rawTokens = (rawTopic.toLowerCase().trim().match(/[a-z0-9]+/g) ?? []);
  let topicDistinct = new Set(rawTokens.filter((t) => t.length > 2 && !GENERIC_TOKENS.has(t) && !MATCH_STOPWORDS.has(t)));
  if (!topicDistinct.size) topicDistinct = new Set(rawTokens.filter((t) => t.length > 1));

  const activeNegatives: string[] = [];
  for (const neg of [...HARD_NEGATIVES, ..._DRIFT_NEGATIVES]) {
    const negTokens = neg.toLowerCase().match(/[a-z0-9]+/g) ?? [];
    const negDistinct = negTokens.filter((t) => t.length > 2 && !GENERIC_TOKENS.has(t) && !MATCH_STOPWORDS.has(t));
    let collision = false;
    for (const td of topicDistinct) {
      for (const nt of negDistinct) {
        if (td === nt || (td.length >= 3 && (td.startsWith(nt) || nt.startsWith(td)))) {
          collision = true;
          break;
        }
      }
      if (collision) break;
    }
    if (!collision) activeNegatives.push(neg);
  }
  const negatives = activeNegatives.length ? activeNegatives : ["sorority", "summer camp", "payroll software"];

  let contrastiveQ: number[];
  try {
    const qPrefix = "Represent this sentence for searching relevant passages: ";
    let respNeg = await postEmbed([qPrefix + rawTopic, ...negatives], 60000);
    if (respNeg.status === 503) {
      log("▸ System warming up — retrying in 15 seconds...");
      await sleep(15000);
      respNeg = await postEmbed([qPrefix + rawTopic, ...negatives], 60000);
    }
    if (respNeg.status !== 200) {
      const body = (await respNeg.text()).slice(0, 200);
      throw new HFAPIError(`Pass 1 embedding returned HTTP ${respNeg.status}: ${body}`);
    }

    const embs: number[][] = await respNeg.json();
    const dim = embs[0].length;
    const negMean = new Array(dim).fill(0);
    for (let i = 1; i < embs.length; i++) for (let d = 0; d < dim; d++) negMean[d] += embs[i][d];
    for (let d = 0; d < dim; d++) negMean[d] /= (embs.length - 1);

    const q = embs[0].map((v, d) => v - CONTRASTIVE_WEIGHT * negMean[d]);
    contrastiveQ = l2normalize(q);
  } catch (exc) {
    if (exc instanceof HFAPIError) throw exc;
    throw new HFAPIError(`Relevance scoring failed: ${exc instanceof Error ? exc.message : String(exc)}`);
  }

  // ── Batched corpus embedding ───────────────────────────────────────────────
  let finalScores: number[];
  try {
    const batches: string[][] = [];
    for (let i = 0; i < corpus.length; i += NLP_BATCH_SIZE) batches.push(corpus.slice(i, i + NLP_BATCH_SIZE));

    const batchResults = await mapWithConcurrency(batches, NLP_EMBED_WORKERS, async (batch) => {
      const r = await postEmbed(batch, 90000);
      if (r.status !== 200) {
        const body = (await r.text()).slice(0, 200);
        throw new HFAPIError(`Batch embedding failed HTTP ${r.status}: ${body}`);
      }
      return (await r.json()) as number[][];
    });

    const mat: number[][] = [];
    for (const b of batchResults) mat.push(...b);
    finalScores = cosineToQuery(contrastiveQ, mat);
  } catch (exc) {
    if (exc instanceof HFAPIError) throw exc;
    throw new HFAPIError(`Relevance scoring failed: ${exc instanceof Error ? exc.message : String(exc)}`);
  }

  // ── Query-only precedence penalty + zero-out floor ─────────────────────────
  return rows.map((r, i) => {
    const threshold = r.matched_on === "Query" ? QUERY_ONLY_MIN_THRESHOLD : RELEVANCE_MIN_THRESHOLD;
    const score = clampRound1(finalScores[i] * 10, SCORE_MIN, SCORE_MAX);
    return score < threshold ? 0.0 : score;
  });
}

File: backend/lib/relevance.py
"""
relevance.py — HF bi-encoder relevance scoring. Port of relevance.ts.
D1: NO TF-IDF fallback. HFAPIError is raised on any failure.
"""
from __future__ import annotations
import asyncio
import re
import httpx
from typing import Optional

from .config import (
    HF_API_TOKEN, HF_EMBED_MODEL, HARD_NEGATIVES, _DRIFT_NEGATIVES,
    CONTRASTIVE_WEIGHT, NLP_BATCH_SIZE, NLP_EMBED_WORKERS,
    RELEVANCE_MIN_THRESHOLD, QUERY_ONLY_MIN_THRESHOLD, SCORE_MIN, SCORE_MAX,
)
from .util import CandidateRow, LogFn, l2normalize, cosine_to_query, clamp_round1, map_with_concurrency

EMBED_URL = f"https://router.huggingface.co/hf-inference/models/{HF_EMBED_MODEL}/pipeline/feature-extraction"

GENERIC_TOKENS = {
    "software", "platform", "platforms", "tool", "tools", "service", "services",
    "solution", "solutions", "system", "systems", "app", "apps", "online",
    "best", "top", "company", "companies", "provider", "providers", "management",
}
MATCH_STOPWORDS = {"the", "and", "for", "with", "your", "you", "our", "from", "that", "this", "are"}


class HFAPIError(Exception):
    pass


async def _post_embed(texts: list[str], timeout_s: float) -> list[list[float]]:
    if not HF_API_TOKEN or not HF_API_TOKEN.startswith("hf_"):
        raise HFAPIError("HF_API_TOKEN is missing or invalid.")
    async with httpx.AsyncClient(timeout=timeout_s) as client:
        resp = await client.post(
            EMBED_URL,
            headers={"Authorization": f"Bearer {HF_API_TOKEN}", "Content-Type": "application/json"},
            json={"inputs": texts},
        )
        if resp.status_code != 200:
            body = resp.text[:200]
            raise HFAPIError(f"Embedding returned HTTP {resp.status_code}: {body}")
        return resp.json()


async def warmup_embed_model() -> None:
    try:
        await _post_embed(["warmup"], 30)
    except Exception:
        pass


def _build_nlp_corpus(rows: list[CandidateRow], raw_topic: str) -> list[str]:
    topic_terms = [t for t in re.findall(r"[a-z0-9]+", raw_topic.lower()) if len(t) > 2]
    corpus = []
    for r in rows:
        bits = [
            str(r.query or r.query_all or ""),
            str(r.meta_title or ""),
            str(r.h1 or ""),
        ]
        for col in [r.h2 or "", r.meta_description or ""]:
            val = str(col)
            if not val:
                continue
            fragments = [f.strip() for f in re.split(r"[|•♦\n.\-]", val) if f.strip()]
            for frag in fragments:
                if any(term in frag.lower() for term in topic_terms):
                    bits.append(frag)
        corpus.append(" ".join(bits))
    return corpus


async def compute_relevance_scores(
    rows: list[CandidateRow],
    raw_topic: str,
    log: LogFn = print,
) -> list[float]:
    if not rows:
        return []
    if not HF_API_TOKEN or not HF_API_TOKEN.startswith("hf_"):
        raise HFAPIError("HF_API_TOKEN is missing or invalid.")

    corpus = _build_nlp_corpus(rows, raw_topic)

    # ── Dynamic collision check (D9-adjacent) ─────────────────────────────────
    raw_tokens = re.findall(r"[a-z0-9]+", raw_topic.lower().strip())
    topic_distinct = {t for t in raw_tokens if len(t) > 2 and t not in GENERIC_TOKENS and t not in MATCH_STOPWORDS}
    if not topic_distinct:
        topic_distinct = {t for t in raw_tokens if len(t) > 1}

    active_negatives: list[str] = []
    for neg in HARD_NEGATIVES + _DRIFT_NEGATIVES:
        neg_tokens = re.findall(r"[a-z0-9]+", neg.lower())
        neg_distinct = [t for t in neg_tokens if len(t) > 2 and t not in GENERIC_TOKENS and t not in MATCH_STOPWORDS]
        collision = any(
            td == nt or (len(td) >= 3 and (td.startswith(nt) or nt.startswith(td)))
            for td in topic_distinct for nt in neg_distinct
        )
        if not collision:
            active_negatives.append(neg)
    negatives = active_negatives if active_negatives else ["sorority", "summer camp", "payroll software"]

    # ── Contrastive query embedding ───────────────────────────────────────────
    try:
        q_prefix = "Represent this sentence for searching relevant passages: "
        try:
            embs = await _post_embed([q_prefix + raw_topic] + negatives, 60)
        except HFAPIError as e:
            if "503" in str(e):
                log("▸ System warming up — retrying in 15 seconds...")
                await asyncio.sleep(15)
                embs = await _post_embed([q_prefix + raw_topic] + negatives, 60)
            else:
                raise

        dim = len(embs[0])
        neg_mean = [0.0] * dim
        for emb in embs[1:]:
            for d in range(dim):
                neg_mean[d] += emb[d]
        for d in range(dim):
            neg_mean[d] /= len(embs) - 1

        q = [embs[0][d] - CONTRASTIVE_WEIGHT * neg_mean[d] for d in range(dim)]
        contrastive_q = l2normalize(q)
    except HFAPIError:
        raise
    except Exception as exc:
        raise HFAPIError(f"Relevance scoring failed: {exc}") from exc

    # ── Batched corpus embedding ──────────────────────────────────────────────
    try:
        batches = [corpus[i:i + NLP_BATCH_SIZE] for i in range(0, len(corpus), NLP_BATCH_SIZE)]

        async def embed_batch(batch: list[str]) -> list[list[float]]:
            return await _post_embed(batch, 90)

        batch_results = await map_with_concurrency(batches, NLP_EMBED_WORKERS, embed_batch)
        mat: list[list[float]] = []
        for b in batch_results:
            mat.extend(b)

        raw_scores = cosine_to_query(contrastive_q, mat)
    except HFAPIError:
        raise
    except Exception as exc:
        raise HFAPIError(f"Relevance scoring failed: {exc}") from exc

    # ── Query-only precedence penalty + zero-out floor ────────────────────────
    return [
        (lambda threshold, score: 0.0 if score < threshold else score)(
            QUERY_ONLY_MIN_THRESHOLD if rows[i].matched_on == "Query" else RELEVANCE_MIN_THRESHOLD,
            clamp_round1(raw_scores[i] * 10, SCORE_MIN, SCORE_MAX),
        )
        for i in range(len(rows))
    ]
